### PR TITLE
Replace Spans by SourcePositions.

### PR DESCRIPTION
**Pull request by sjrd**
Unlike spans, SourcePositions keep a reference to their SourceFile. The latter stores the sizes of its lines, which allows it to reconstruct line and column numbers from offsets.

### CODE REVIEW

File: tasty-query/shared/src/main/scala/tastyquery/SourceFile.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    else throw UnsupportedOperationException(s"Source file $path does not have line/column information")
+
+  private def lineToOffset(line: Int): Int =
+    ensureLineStartOffsets()(line)
+
+  private[tastyquery] def offsetToLine(offset: Int): Int =
+    val startOffsets = ensureLineStartOffsets()
+    val p = Arrays.binarySearch(startOffsets, offset)
+    if p >= 0 then p
+    else (-p - 1) - 1 // (-p - 1) is the insertion point, from which we still need to subtract 1 to get the line
```
Comment (adpi2) [inline tasty-query/shared/src/main/scala/tastyquery/SourceFile.scala:42]: That's funky!

File: tasty-query/shared/src/main/scala/tastyquery/SourcePosition.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  def endOffset: Int = span.end
+
+  /** The line number of the end of this position. */
+  def endLine: Int = sourceFile.offsetToLine(endOffset)
+
+  /** The column number of the end of this position. */
+  def endColumn: Int = sourceFile.offsetToColumn(endOffset)
+
+  /** Is this position synthetic? */
+  def isSynthetic: Boolean = span.isSynthetic
```
Comment (adpi2) [inline tasty-query/shared/src/main/scala/tastyquery/SourcePosition.scala:56]: What's a synthetic span?

**Review by adpi2 — APPROVED** [review]
Formidable!

File: tasty-query/shared/src/main/scala/tastyquery/SourcePosition.scala (hunk 2 above)
Comment (sjrd) [reply tasty-query/shared/src/main/scala/tastyquery/SourcePosition.scala:61]: I added a comment. Is that clearer?

**Pull request merged**